### ISSUE TITLE
Sort by multiple columns of different types

### ISSUE THREAD

**Issue by ermalguni** [vaexio/vaex#815]
Is it possible to sort by multiple columns where the column data types are ints and strings?

**Comment by JovanVeljanoski** [comment]
Yes. did you try it?

**Comment by ermalguni** [comment]
Yes, and I got this: `TypeError: '<' not supported between instances of 'str' and 'float'`.

**Comment by JovanVeljanoski** [comment]
You asked for ints right:) The error suggests that you are attempting to sort floats and strings. 

Joking aside, the sort of vaex should follow numpy convention. I just tried it out with ints and strings, and it works. I actually don't know why it would.. i was thought it would not. Maybe @maartenbreddels would have a better insight into this. 

In general, when talking about sorting strings, you generally refer to sorting them alphabetically right?

**Comment by ermalguni** [comment]
Yes, I asked about ints because that is what the data types are, but the exception message talks about float32. That got me confused as to where did the float32 come from.

Regarding sorting string, yes I mean alphabetical sorting. My data contains geographical data.

**Comment by maartenbreddels** [comment]
Could you give us a piece of code to reproduce the issue? (e.g. using vaex.from_arrays(...)`

**Comment by abaradia1978** [comment]
<img width="537" alt="vaex_sorting_str_error" src="https://user-images.githubusercontent.com/33593180/87909646-42996900-ca86-11ea-8e73-a0c4a8e3a2b2.png">

Attached screenshot shows the error that appears while sorting string fields - the same works find with pandas. let me know if you need any clarification

**Comment by abaradia1978** [comment]
<img width="346" alt="vaex_sorting_ascdes_error" src="https://user-images.githubusercontent.com/33593180/87910051-fd296b80-ca86-11ea-888b-18f208b9d1d4.png">

Also, while sorting multiple fields, if we give different sort order, it does not work - refer above screen shot - it works as expected in pandas

**Comment by JovanVeljanoski** [comment]
Hi @abaradia1978 

Specifying distinct ordering per sorting key is not yet supported. We will discuss this if we can implement it efficiently. 
Thanks for bringing this to our attentino.

**Comment by abaradia1978** [comment]
Thanks @JovanVeljanoski - we currently stuck because of this issue in one of our projects. Look forward to having it resolved. Also will you consider allowing sort on string fields as per my first screen shot above?

**Comment by JovanVeljanoski** [comment]
Hi @abaradia1978  et al, 

I just tried out the sorting stuff, and it looks like sorting string columns does work.  Which numpy version do you have? I am running 1.17.5.

Btw: screenshots may be convenient for you, but it is a pain for me to re-write your example, so here is my own:
```
x = np.array([5, 3, 1, 1, 5])
y = np.array([0, 3, 4, 2, 1])
z = np.array(['dog', 'cat', 'cat', 'dog', 'mouse'])
df = vaex.from_arrays(x=x, y=y, z=z)

df.sort(['z', 'y'])  # And the output is correctly printed. 
```

As for the multi order thing - how would you do it with numpy?

**Comment by abaradia1978** [comment]
Hi @JovanVeljanoski ,

I don't think numpy version is an issue. Not using nummy array at all is the issue.

In my code, I was simply making dataframe as under.

`df = pd.DataFrame({
    'col1': ['A', 'A', 'B', np.nan, 'D', 'C'],
    'col2': [2, 2, 3, 3, 7, 4],
    'col3': [0, 1, 9, 4, 2, 3],
})`

But it worked if I wrapped it around nummy array as under

`df = pd.DataFrame({
    'col1': np.array(['A', 'A', 'B', 'B', 'D', 'C']),
    'col2': np.array([2, 2, 3, 3, 7, 4]),
    'col3': np.array([0, 1, 9, 4, 2, 3]),
})`

But sorting multiple fields is still an issue - for nummy array providing sort order as array as shown below works but vaex df simply ignores the array and sort by ascending order by default.

`print(df.sort_values(by=['col1','col3'],ascending=[False,True])) # this works`
`print(vx.sort(by=['col1','col3'],ascending=[False])) # this sorts ascending by default`

And there is an another issue. If we make a vaex dataframe from csv or HD5 file, it won't be possible to convert string columns to np array every time. Check out below code. It only sorts in ascending order even if the second parameter to sort function is False (3rd print statement).

`import vaex`
`df = vaex.from_csv('test.csv',header=0)` 
`print(df)`
`print(df.sort(by=['col1']))`
`print(df.sort(by=['col1'],ascending='False'))`

output is as under

![image](https://user-images.githubusercontent.com/33593180/88536341-4c444300-d029-11ea-8bbc-849124b35f51.png)

my csv file contents

```
col1,col2,col3
A,2,0
A,2,1
B,3,9
B,3,4
D,7,2
C,4,3
```

**Comment by JovanVeljanoski** [comment]
Hi @abaradia1978 

You are raising different, not really related points, but I will try to answer them in turn:

1) 
In your example, you a creating a vaex dataframe from a pandas dataframe, in both of your cases. Vaex does not always play nice when a pandas column (series) is of dtype object. When you create a pandas dataframe out of numpy arrays, it is numpy that does something to the list before it is passed to pandas. In general you want all rows of a columns to have the same type. The dtype object will likely disappear in the future. 

So you can either handle this on the pandas side (i.e. impute the nans, drop rows etc..), or do the same on the vaex side. 
Note that you don't have to use a pandas dataframe to create a vaex dataframe. If you have a dict, you can already do:
`ds = vaex.from_dict({'col1': ['A', 'A', 'B', np.nan, 'D', 'C'], 'col2': [2, 2, 3, 3, 7, 4], 'col3': [0, 1, 9, 4, 2, 3]})`.

In this case, the sorting on `col1` will work. 

2) 
As i said in the earlier post, providing distinct ordering for multiple columns is currently not supported. So you can sort by multiple columns, but you can't choose an ordering for each key separately: either everything is ascending, or descending.

3)
For reading of the CSV.. i can't reproduce the issue, i tried it on Linux and Mac and I get it to work correctly in both cases:
<img width="1213" alt="image" src="https://user-images.githubusercontent.com/18574951/88573427-04311a80-d041-11ea-8df0-a046a440f2fc.png">

**Comment by abaradia1978** [comment]
Thanks @JovanVeljanoski for your reply. You are correct, there is no issue with point 3 - sorting with csv is working fine even for string columns. My bad - I was passing False in quotes. :-( 

look forward to having multiple sort order point resolved.
